flink-runtime/src/main/java/org/apache/flink/runtime/rpc/resourcemanager/ResourceManagerToTaskExecutorHeartbeatScheduler.java

```diff
@@ -0,0 +1,242 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.rpc.resourcemanager;
+
+import akka.dispatch.OnFailure;
+import akka.dispatch.OnSuccess;
+import org.apache.flink.runtime.clusterframework.types.ResourceID;
+import org.apache.flink.runtime.rpc.taskexecutor.SlotReport;
+import org.apache.flink.runtime.rpc.taskexecutor.TaskExecutorGateway;
+import org.slf4j.Logger;
+import scala.concurrent.Future;
+import scala.concurrent.duration.FiniteDuration;
+
+import java.util.UUID;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import static org.apache.flink.util.Preconditions.checkState;
+
+/**
+ * heartbeat between ResourceManager and TaskManager, it is responsible for schedule heartbeat and handle
+ * heartbeat lost cases
+ */
+public class ResourceManagerToTaskExecutorHeartbeatScheduler {
+
+	/** default heartbeat interval time in millisecond */
+	private static final long INITIAL_HEARTBEAT_INTERVAL_MILLIS = 5000;
+
+	/** default heartbeat timeout in millisecond */
+	private static final long INITIAL_HEARTBEAT_TIMEOUT_MILLIS = 200;
+
+	/** max heartbeat interval time in millisecond (which is used in retry heartbeat case) */
+	private static final long MAX_HEARTBEAT_TIMEOUT_MILLIS = 30000;
+
+	/** if a failure except for timeout exception happened when trigger heartbeat from resourceManager to taskManager , next attemp will start after  ERROR_HEARTBEAT_DELAY_MILLIS millisecond */
+	private static final long ERROR_HEARTBEAT_DELAY_MILLIS = 2000;
+
+	/** max heartbeat retry times when lost heartbeat */
+	private static final int MAX_ATTEMPT_TIMES = 8;
+
+	private final long heartbeatInterval;
+
+	private final long heartbeatTimeout;
+
+	private final long maxHeartbeatTimeout;
+
+	private final long delayOnError;
+
+	private final int maxAttempt;
+
+
+	/** taskManagerGateway to receive the heartbeat and report slot allocation */
+	private final TaskExecutorGateway taskExecutorGateway;
+
+	/** the taskManager address */
+	private final String taskExecutorAddress;
+
+	/** identify the taskManager resourceID */
+	private final ResourceID resourceID;
+
+	/** identify the resourceManager rpc endpoint */
+	private final ResourceManager resourceManager;
+
+	private final UUID resourceManagerLeaderSessionID;
+
+	private final Logger log;
+
+	private volatile boolean closed;
+
+	/**
+	 * ResourceManagerToTaskExecutorHeartbeatScheduler constructor
+	 *
+	 * @param resourceManager         resourceManager which handles heartbeat communication with taskManager
+	 * @param taskExecutorGateway     taskManager which receives heartbeat from resourceManager and report its slot
+	 *                                allocation to resourceManager
+	 * @param taskExecutorAddress     taskManager's address
+	 * @param taskExecutorResourceID  taskManager's resourceID
+	 * @param log
+	 */
+	public ResourceManagerToTaskExecutorHeartbeatScheduler(
```

Comment:
I think we also need a receiving end of the heartbeat because the `TaskExecutor` wants to monitor the `ResourceManager` as well. If the heartbeat request is not delivered, then the RM should be marked as dead.